gap/lqueue.gi

```diff
@@ -1,174 +1,203 @@
-      tail := tail + 1;
-      queue![QTAIL] := tail;
+function(queue,el)
+    local head, tail, last;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # Special case for empty queue,
+    # head = tail and queue![QDATA][tail] is
+    # not bound
+    if not IsBound(queue![QDATA][tail]) then
+        queue![QDATA][tail] := el;
     else
-      queue![QTAIL] := 1;
-      queue![QDATA][last] := el;
+        if tail = last then
+            tail := 1;
+        else
+            tail := tail + 1;
+        fi;
+        queue![QDATA][tail] := el;
+        queue![QTAIL] := tail;
+
+        # If queue is full expand
+        if ((head = 1) and (tail = last))
+           or (tail + 1 = head) then
+            PlistQueueExpand(queue);
+        fi;
     fi;
-  elif tail + 1 <> head then
-    queue![QDATA][tail] := el;
-    tail := tail + 1;
-    queue![QTAIL] := tail;
-  else
-    PlistQueueExpand(queue);
-    tail := queue![QTAIL];
-    queue![QDATA][tail] := el;
-    tail := tail + 1;
-    queue![QTAIL] := tail;
-  fi;
 end);
 
 InstallGlobalFunction(PlistQueuePushFront,
 function(queue, el)
-  local head, tail, last;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if head = 1 then
-    if tail = last then
-      PlistQueueExpand(queue);
-      head := queue![QCAPACITY];
-      queue![QDATA][head] := el;
-      queue![QHEAD] := head;
+    local head, tail, last;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # If the queue is empty
+    if not IsBound(queue![QDATA][head]) then
+        queue![QDATA][head] := el;
     else
-      queue![QHEAD] := last;
-      queue![QDATA][last] := el;
+        if head = 1 then
+            head := last;
+        else
+            head := head - 1;
+        fi;
+        queue![QDATA][head] := el;
+        queue![QHEAD] := head;
+
+        if (head = 1 and tail = last)
+           or (tail + 1 = head) then
+            PlistQueueExpand(queue);
+        fi;
     fi;
-  elif tail + 1 <> head then
-    head := head - 1;
-    queue![QDATA][head] := el;
-    queue![QHEAD] := head;
-  else
-    PlistQueueExpand(queue);
-    head := Size(queue);
-    queue![QDATA][head] := el;
-    queue![QHEAD] := head;
-  fi;
 end);
 
 InstallGlobalFunction(PlistQueuePopFront,
 function(queue)
-  local head, tail, last, result;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if head <> tail then
-    if head = last then
-      head := 1;
-      result := queue![QDATA][last];
-      queue![QDATA][last] := fail;
-    else
-      head := head + 1;
-      result := queue![QDATA][head-1];
-      queue![QDATA][head-1] := fail;
+    local head, tail, last, result;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # TODO: What if someone wants to store fails?
+    result := fail;
+
+    if IsBound(queue![QDATA][head]) then
```

Comment:
Personally, I'd find it easier to read this code if it did:
```gap
  if not IsBound(queue![QDATA][head]) then
    return fail;
  fi;
  ...
  return result;
```
The same applies other similar locations, too, of course.